lib/services/discourse.ts

```diff
@@ -1,44 +1,138 @@
-    private topicCache = new Map<string, any>();
+    // There are circumstances in which the discourse web-hook will fire twice for the same post, so track.
     private postsSynced = new Set<number>();
 
     /**
-     * Retrieve the comments in a thread that match an optional filter
-     * @param event details to identify the event
-     * @param filter regex of comments to match
+     * Promise to turn the data enqueued into a generic message format
+     * @param data - Raw data from the enqueue, remembering this is as dumb and quick as possible
+     * @returns {Bluebird<ReceiptContext>} - A promise that resolves to the generic form of the event
      */
-    public fetchThread(event: ReceiptContext, filter: RegExp): Promise<string[]> {
-        // Check that the event being asked about orginated with us
-        if (event.source !== this.serviceName) {
-            return Promise.reject(new Error('Cannot get discourse thread from non-discourse event'));
+    public makeGeneric(data: MessengerEvent): Promise<ReceiptContext> {
+        // Encode once the common parts of a request
+        const getGeneric = {
+            json: true,
+            method: 'GET',
+            qs: {
+                api_key: process.env.DISCOURSE_LISTENER_ACCOUNT_API_TOKEN,
+                api_username: process.env.DISCOURSE_LISTENER_ACCOUNT_USERNAME,
+            },
+            // appended before execution
+            uri: `https://${process.env.DISCOURSE_INSTANCE_URL}`,
+        };
+        // Gather more complete details of the enqueued event
+        const getPost = _.cloneDeep(getGeneric);
+        getPost.uri += `/posts/${data.rawEvent.id}`;
+        const getTopic = _.cloneDeep(getGeneric);
+        getTopic.uri += `/t/${data.rawEvent.topic_id}`;
+        return Promise.props({
+            post: request(getPost),
+            topic: request(getTopic),
+        })
+        .then((details: {post: any, topic: any}) => {
+            // Gather metadata and resolve
+            const metadata = MessageService.extractMetadata(details.post.raw);
+            const first = details.post.post_number === 1;
+            return {
+                action: 'create',
+                first,
+                genesis: metadata.genesis || data.source,
+                // post_type 4 seems to correspond to whisper
+                hidden: first ? !details.topic.visible : details.post.post_type === 4,
+                source: 'discourse',
+                sourceIds: {
+                    // These come in as integers, but should be strings
+                    flow: details.topic.category_id.toString(),
+                    message: details.post.id.toString(),
+                    thread: details.post.topic_id.toString(),
+                    url: getTopic.uri,
+                    user: details.post.username,
+                },
+                text: metadata.content,
+                title: details.topic.title,
+            } as ReceiptContext;
```

Comment:
Note to self: To discuss.
Types of property 'action' are incompatible.  Type 'string' is not assignable to type '"create"'.